builtin/joined_dataset.cc

```diff
@@ -360,6 +239,223 @@ struct JoinedDataset::Itl
         }
     }
 
+     /* This is called to record a new entry from the join. */
+    void recordJoinRow(const RowName & leftName, RowHash leftHash, const RowName & rightName, RowHash rightHash)
+    {
+        bool debug = false;
+
+        RowName rowName(leftName.toUtf8String() + "-" + rightName.toUtf8String());
+        RowHash rowHash(rowName);
+
+        RowEntry entry;
+        entry.rowName = rowName;
+        entry.rowHash = rowHash;
+        entry.leftName = leftName;
+        entry.rightName = rightName;
+
+        if (debug)
+            cerr << "added entry number " << rows.size()
+                 << "named " << "("<< rowName.toUtf8String() <<")"
+                 << endl;
+
+        rows.emplace_back(std::move(entry));
+        rowIndex[rowHash] = rows.size() - 1;
+
+        leftRowIndex[leftHash].push_back(rowName);
+        rightRowIndex[rightHash].push_back(rowName);
+    };
+
+    void makeJoinFast(AnnotatedJoinCondition& condition, SqlExpressionMldbContext& context, BoundTableExpression& left, BoundTableExpression& right, JoinQualification qualification)
+    {
+        bool debug = false;
+        bool outerLeft = qualification == JOIN_LEFT || qualification == JOIN_FULL;
+        bool outerRight = qualification == JOIN_RIGHT || qualification == JOIN_FULL;
+
+        // Where expressions for the left and right side
+        auto runSide = [&] (const AnnotatedJoinCondition::Side & side,
+                            const Dataset & dataset,
+                            bool outer,
+                            const std::function<void (const RowName&, const RowHash& )> & recordOuterRow)
+            -> std::vector<std::tuple<ExpressionValue, RowName, RowHash> >
+            {
+                auto condition = side.where;
+
+                std::vector<std::shared_ptr<SqlExpression> > clauses = { side.selectExpression };
+
+                if (outer)
+                {
+                    //return all rows
+                    condition = SqlExpression::TRUE;
```

Comment:
ugh, I renamed a parameter with a local, I'll change that. The crossWhere are the non-pivot conditions, but as I understand each side of the pivot can still have conditions in the fast path.